### PR TITLE
[WIP] add TF-style julia benchmarks, re-organize test/perf code, and other stuff

### PR DESCRIPTION
**Pull request by jrevels**
The most recent commit on this branch is in a half-finished state, but I figured I'd open the PR just to get eyes on the changes here

### CODE REVIEW

File: src/MixedModeBroadcastAD.jl [hunk 1]
```diff
@@ -13,48 +13,41 @@ using CUDAnative
...
 
-macro cuda_linear_index(A)
-    esc(quote
-        i = (blockIdx().x-UInt32(1)) * blockDim().x + threadIdx().x
-        i > length($A) && return
-        i
-    end)
+@inline function cuda_linear_index(n::Integer)
+    i = (blockIdx().x - UInt32(1)) * blockDim().x + threadIdx().x
+    i > n && return
```
Comment (maleadt) [inline src/MixedModeBroadcastAD.jl]: This needs to be a macro to cause an exit of the parent function (a design decision from the pre-bitstype union era, we could probably get away returning `nothing`).

Comment (jrevels) [reply src/MixedModeBroadcastAD.jl]: Oh, okay. Thanks.

**Comment by jrevels** [comment]
Okay, tests finally pass here with the TF-style kernels. All that's left now is to finish shuffling around some of the profile/benchmark code and then we should be good to go.

**Comment by maleadt** [comment]
FWIW, `prod(length, tuple)` doesn't work because of missing inbounds handling & propagation in Base:

```patch
diff --git a/base/reduce.jl b/base/reduce.jl
index 12b231bda6..3a67c23966 100644
--- a/base/reduce.jl
+++ b/base/reduce.jl
@@ -40,7 +40,7 @@ mul_prod(x::SmallUnsigned,y::SmallUnsigned) = UInt(x) * UInt(y)
     if done(itr, i)
         return v0
     else
-        (x, i) = next(itr, i)
+        @inbounds (x, i) = next(itr, i)
         v = op(v0, f(x))
         while !done(itr, i)
             @inbounds (x, i) = next(itr, i)
@@ -71,7 +71,7 @@ function mapfoldl(f, op, itr)
     if done(itr, i)
         return Base.mapreduce_empty_iter(f, op, itr, IteratorEltype(itr))
     end
-    (x, i) = next(itr, i)
+    @inbounds (x, i) = next(itr, i)
     v0 = mapreduce_first(f, op, x)
     mapfoldl_impl(f, op, v0, itr, i)
 end
diff --git a/base/tuple.jl b/base/tuple.jl
index f85c53d8a4..0b5eb71d3c 100644
--- a/base/tuple.jl
+++ b/base/tuple.jl
@@ -37,7 +37,7 @@ _setindex(v, i::Integer) = ()
 
 start(t::Tuple) = 1
 done(t::Tuple, i::Int) = (length(t) < i)
-next(t::Tuple, i::Int) = (t[i], i+1)
+next(t::Tuple, i::Int) = (@_propagate_inbounds_meta; (t[i], i+1))
 
 keys(t::Tuple) = 1:length(t)
```

But the GPU-compatible alternative is also 10% faster (`prod(f, x)` relies on `@noinline mapfoldl`).
Pretty sad to require this kind of complexity for computing an index.

**Pull request merged**